app_integrations/batcher.py

```diff
@@ -0,0 +1,122 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import json
+import math
+
+import boto3
+from botocore.exceptions import ClientError
+
+from app_integrations import LOGGER
+
+# Max lambda input payload size is 128K for the 'event' invocation type
+MAX_LAMBDA_PAYLOAD_SIZE = 128 * 1000
+
+
+class Batcher(object):
+    """Batcher class to handle segmenting large messages going to the
+    rule processor lambda function
+    """
+    LAMBDA_CLIENT = None
+
+    def __init__(self, config):
+        # Create the lambda client if it does not exist
+        if Batcher.LAMBDA_CLIENT is None:
+            Batcher.LAMBDA_CLIENT = boto3.client('lambda', region_name=config['region'])
+        # The rule processor function name will look like:
+        # <prefix>_<cluster>_streamalert_rule_processor
+        self.rp_function = '_'.join([config['prefix'], config['cluster'],
+                                     'streamalert_rule_processor'])
+
+    def send_logs(self, source_function, logs):
+        """Public method to send the logs to the rule processor
+
+        Args:
+            source_function (str): The app function name from which the logs came
+            logs (list): List of the logs that have been gathered
+        """
+        LOGGER.info('Starting batch send of %d logs to the rule processor', len(logs))
+
+        # Try to send all of the logs in one fell swoop
+        if self._send_logs_to_stream_alert(source_function, logs):
+            return
+
+        # Fall back on segmenting the list of logs into multiple requests
+        # if they could not be sent at once
+        return self._segement_and_send(source_function, logs)
+
+    def _segement_and_send(self, source_function, logs):
+        """Protected method for segmenting a list of logs into smaller lists
+        so they conform to the input limit of AWS Lambda
+
+        Args:
+            source_function (str): The app function name from which the logs came
+            logs (list): List of the logs that have been gathered
+        """
+        log_count = len(logs)
+        LOGGER.debug('Segmenting %d logs into subsets', log_count)
+
+        segment_size = int(math.ceil(log_count / 2.0))
+        for index in range(0, log_count, segment_size):
+            subset = logs[index:segment_size + index]
+            # Try to send this current subset to the rule processor
+            # and segment again if they are too large to be sent at once
+            if not self._send_logs_to_stream_alert(source_function, subset):
```

Comment:
If a log exceeds the size of `MAX_LAMBDA_PAYLOAD_SIZE`, `_send_logs_to_stream_alert` will return False. `not` of `False` is True, so `_segement_and_send` will be called, and that function doesn't check MAX_LAMBDA_PAYLOAD_SIZE?

Reply:
The call to`_segement_and_send` is this a recursive call to this same function (which then splits the list and calls `_send_logs_to_stream_alert` again with smaller lists where the size is checked). Let me know if it needs more explanation. See `test_batcher.py` for tests related to this.